Objective-C/CBLDatabase.mm

```diff
@@ -213,12 +213,69 @@ - (BOOL) saveDocument: (CBLMutableDocument*)document
     CBLAssertNotNil(document);
     
     return [self saveDocument: document
+             withBaseDocument: nil
            concurrencyControl: concurrencyControl
                    asDeletion: NO
                         error: error];
 }
 
 
+- (BOOL) saveDocument: (CBLMutableDocument *)document
+      conflictHandler: (BOOL (^)(CBLMutableDocument *, CBLDocument * nullable))conflictHandler
+                error: (NSError **)error {
+    
+    CBLAssertNotNil(document);
+    CBLAssertNotNil(conflictHandler);
+    
+    while (true) {
+        NSError* err;
+        CBLDocument* oldDoc = nil;
+        BOOL success = [self saveDocument: document
+                         withBaseDocument: oldDoc
+                       concurrencyControl: kCBLConcurrencyControlFailOnConflict
+                               asDeletion: NO
+                                    error: &err];
+        // if it's a conflict, we will use the conflictHandler to resolve.
+        if (!success && $equal(err.domain, CBLErrorDomain) && err.code == CBLErrorConflict) {
+            CBL_LOCK(self) {
+                C4Transaction transaction(_c4db);
+                if (!transaction.begin())
+                    return convertError(transaction.error(), error);
+                
+                oldDoc = [[CBLDocument alloc] initWithDatabase: self
+                                                    documentID: document.id
+                                                includeDeleted: YES
+                                                         error: error];
+                if (!oldDoc)
+                    return NO;
+                
+                if (!transaction.commit())
+                    return convertError(transaction.error(), error);
+            }
+            
+            @try {
+                if (conflictHandler(document, oldDoc)) {
```

Comment:
We also need to check if oldDoc is a deleted doc or not. If it's a deleted doc, the null value should be given to conflictHandler.

Reply:
done